pkg/controller/metrics/provision_underway_collector.go

```diff
@@ -96,14 +105,132 @@ func (cc provisioningUnderwayCollector) Describe(ch chan<- *prometheus.Desc) {
+		"hive_cluster_deployment_provision_underway_seconds",
+		"Length of time a cluster has been provisioning.",
+		[]string{"cluster_deployment", "namespace", "cluster_type", "condition", "reason", "platform", "image_set"},
+		nil,
+	)
+)
+
 func newProvisioningUnderwayCollector(client client.Client) prometheus.Collector {
 	return provisioningUnderwayCollector{
 		client: client,
-		metricClusterDeploymentProvisionUnderwaySeconds: prometheus.NewDesc(
-			"hive_cluster_deployment_provision_underway_seconds",
-			"Length of time a cluster has been provisioning.",
-			[]string{"cluster_deployment", "namespace", "cluster_type", "condition", "reason", "platform", "image_set"},
-			nil,
-		),
+		metricClusterDeploymentProvisionUnderwaySeconds: metricClusterDeploymentProvisionUnderwaySecondsDesc,
+	}
+}
+
+func newProvisioningUnderwayCollectorWithMinDuration(client client.Client, minimum time.Duration) prometheus.Collector {
+	return provisioningUnderwayCollector{
+		client: client,
+		metricClusterDeploymentProvisionUnderwaySeconds: metricClusterDeploymentProvisionUnderwaySecondsDesc,
+		minDuration: minimum,
+	}
+}
+
+// provisioning underway install restarts metrics collected through a custom prometheus collector
+type provisioningUnderwayInstallRestartsCollector struct {
+	client client.Client
+
+	// minRestarts is the minimum restarts after which clusters provisioning
+	// will start reporting the metric. Only used when non-zero.
+	minRestarts int
+
+	// metricClusterDeploymentProvisionUnderwayInstallRestarts is a prometheus metric for the number of install
+	// restarts for a still provisioning cluster.
+	metricClusterDeploymentProvisionUnderwayInstallRestarts *prometheus.Desc
+}
+
+// collects the metrics for provisioningUnderwayCollector
+func (cc provisioningUnderwayInstallRestartsCollector) Collect(ch chan<- prometheus.Metric) {
+	ccLog := log.WithField("controller", "metrics")
+	ccLog.Info("calculating provisioning underway install restarts metrics across all ClusterDeployments")
+
+	// Load all ClusterDeployments so we can accumulate facts about them.
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err := cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+		return
+	}
+	for _, cd := range clusterDeployments.Items {
+		if cd.DeletionTimestamp != nil {
+			continue
+		}
+		if cd.Spec.Installed {
+			continue
+		}
+
+		// Add install failure details for stuck provision
+		condition, reason := "Unknown", "Unknown"
+		for _, delayCondition := range provisioningDelayCondition {
+			if cdCondition := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions,
+				delayCondition); cdCondition != nil {
+				if cdCondition.Status == corev1.ConditionTrue && cdCondition.Reason != "" {
+					condition = string(delayCondition)
+					reason = cdCondition.Reason
+				}
+				break
+			}
+		}
```

Comment:
Probably worth pulling this out to a little function to share between the collectors. Might be subtle fixes in there in future we'd miss in both.

Reply:
updated to keep the condition extraction in a separate function.